File: lib/coroutine/coroutine.go
package coroutine

import (
	"fmt"

	"github.com/arnodel/golua/lib/packagelib"
	rt "github.com/arnodel/golua/runtime"
)

// LibLoader allows loading the coroutine lib
var LibLoader = packagelib.Loader{
	Load: load,
	Name: "coroutine",
}

func load(r *rt.Runtime) (rt.Value, func()) {
	pkg := rt.NewTable()

	rt.SolemnlyDeclareCompliance(
		rt.ComplyCpuSafe|rt.ComplyMemSafe|rt.ComplyTimeSafe|rt.ComplyIoSafe,

		r.SetEnvGoFunc(pkg, "close", close, 1, false), // Lua 5.4
		r.SetEnvGoFunc(pkg, "create", create, 1, false),
		r.SetEnvGoFunc(pkg, "isyieldable", isyieldable, 1, false),
		r.SetEnvGoFunc(pkg, "resume", resume, 1, true),
		r.SetEnvGoFunc(pkg, "running", running, 0, false),
		r.SetEnvGoFunc(pkg, "status", status, 1, false),
		r.SetEnvGoFunc(pkg, "wrap", wrap, 1, false),
		r.SetEnvGoFunc(pkg, "yield", yield, 0, true),
	)

	return rt.TableValue(pkg), nil
}

func close(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	if err := c.Check1Arg(); err != nil {
		return nil, err
	}
	co, err := c.ThreadArg(0)
	if err != nil {
		return nil, err
	}
	ok, err := co.Close(t)
	if !ok {
		return nil, fmt.Errorf("cannot close %s thread", statusString(t, co))
	}
	next := c.Next()
	t.Push1(next, rt.BoolValue(err == nil))
	if err != nil {
		t.Push1(next, rt.ErrorValue(err))
	}
	return next, nil
}

func create(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	var f rt.Callable
	err := c.Check1Arg()
	if err == nil {
		f, err = c.CallableArg(0)
	}
	if err != nil {
		return nil, err
	}
	co := rt.NewThread(t.Runtime)
	co.Start(f)
	return c.PushingNext1(t.Runtime, rt.ThreadValue(co)), nil
}

func resume(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	var co *rt.Thread
	err := c.Check1Arg()
	if err == nil {
		co, err = c.ThreadArg(0)
	}
	if err != nil {
		return nil, err
	}
	res, err := co.Resume(t, c.Etc())
	next := c.Next()
	if err == nil {
		t.Push1(next, rt.BoolValue(true))
		t.Push(next, res...)
	} else {
		t.Push1(next, rt.BoolValue(false))
		t.Push1(next, rt.ErrorValue(err))
	}
	return next, nil
}

func yield(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	res, err := t.Yield(c.Etc())
	if err != nil {
		return nil, err
	}
	return c.PushingNext(t.Runtime, res...), nil
}

func isyieldable(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	var (
		co  = t
		err error
	)
	if c.NArgs() >= 1 {
		co, err = c.ThreadArg(0)
		if err != nil {
			return nil, err
		}
	}
	return c.PushingNext1(t.Runtime, rt.BoolValue(!co.IsMain())), nil
}

func status(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	var co *rt.Thread
	err := c.Check1Arg()
	if err == nil {
		co, err = c.ThreadArg(0)
	}
	if err != nil {
		return nil, err
	}
	return c.PushingNext1(t.Runtime, rt.StringValue(statusString(t, co))), nil
}

func running(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	next := c.Next()
	t.Push1(next, rt.ThreadValue(t))
	t.Push1(next, rt.BoolValue(t.IsMain()))
	return next, nil
}

func wrap(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
	var f rt.Callable
	err := c.Check1Arg()
	if err == nil {
		f, err = c.CallableArg(0)
	}
	if err != nil {
		return nil, err
	}
	co := rt.NewThread(t.Runtime)
	co.Start(f)
	w := rt.NewGoFunction(func(t *rt.Thread, c *rt.GoCont) (rt.Cont, error) {
		res, err := co.Resume(t, c.Etc())
		if err != nil {
			return nil, err
		}
		return c.PushingNext(t.Runtime, res...), nil
	}, "wrap", 0, true)
	w.SolemnlyDeclareCompliance(rt.ComplyCpuSafe | rt.ComplyMemSafe | rt.ComplyTimeSafe | rt.ComplyIoSafe)
	next := c.Next()
	t.Push1(next, rt.FunctionValue(w))
	return next, nil
}

func statusString(running, co *rt.Thread) (status string) {
	if co == running {
		status = "running"
	} else {
		switch co.Status() {
		case rt.ThreadDead:
			status = "dead"
		case rt.ThreadSuspended:
			status = "suspended"
		case rt.ThreadOK:
			status = "normal"
		default:
			status = "unknown"
		}
	}
	return
}

File: lib/coroutine/lua/coroutine.lua
-- Some general checks
do
    local function cof(x)
        print("in cof", x)
        print("in cof", coroutine.yield(x + 2))
        return "from cof"
    end

    local co = coroutine.create(cof)

    print(pcall(coroutine.resume, {}))
    --> ~false\t.*#1 must be a thread

    print("out", coroutine.resume(co, 1))
    print("out", coroutine.resume(co, "two"))
    
    --> =in cof	1
    --> =out	true	3
    --> =in cof	two
    --> =out	true	from cof

    print(coroutine.status(co))
    --> =dead

    print(pcall(coroutine.status, cof))
    --> ~false\t.*#1 must be a thread

    print(pcall(coroutine.create, 1))
    --> ~false\t.*#1 must be a callable
end

-- Check the main coroutine is correctly identified
print(coroutine.running())
--> ~^thread:.*\ttrue$

do
    -- Check the main coroutine is not yieldable
    print(coroutine.isyieldable())
    --> =false

    print(pcall(coroutine.yield, 1))
    --> ~false\t.*cannot yield from main thread

    print(pcall(coroutine.isyieldable, 1))
    --> ~false\t.*must be a thread
end
 
-- Check that coroutine.running() returns true as second argument when
-- called from a non-main coroutine and that a non main coroutine is
-- yieldable.
do
    local function cof()
        print("yieldable", coroutine.isyieldable())
        return coroutine.running()
    end
    local co = coroutine.create(cof)

    print(coroutine.isyieldable(co))
    --> =true

    print(coroutine.resume(co))
    --> =yieldable	true
    --> ~^true\tthread:.*\tfalse$
end

-- Test error in coroutine
do
    local function cof()
        error("boo")
    end
    local co = coroutine.create(cof)
    print(coroutine.resume(co))
    --> ~^false\t.* boo$
    print(coroutine.status(co))
    --> =dead
end

-- Check various statuses of coroutines
do
    local main = coroutine.running()
    local function cof(co)
        coroutine.yield(coroutine.status(co))
        return coroutine.status(main)
    end
    co = coroutine.create(cof)

    print("main/main", coroutine.status(main))
    --> =main/main	running

    print("co/main", coroutine.status(co))
    --> =co/main	suspended

    print("co/co", coroutine.resume(co, co))
    --> =co/co	true	running

    print("main/co", coroutine.resume(co))
    --> =main/co	true	normal

    print("co", coroutine.status(co))
    --> =co	dead
end

-- Test coroutine.wrap()
do
    local function cofib()
        local a, b = 0, 1
        while true do
            coroutine.yield(a)
            a, b = b, a+b
        end
    end
    local fib = coroutine.wrap(cofib)
    print(fib(), fib(), fib(), fib(), fib())
    --> =0	1	1	2	3

    print(pcall(coroutine.wrap))
    --> ~false\t.*value needed

    print(pcall(coroutine.wrap, "not a function"))
    --> ~false\t.*must be a callable

    local werr = coroutine.wrap(function(x, y)
        coroutine.yield(x)
        error(y)
    end)
    print(werr("abc", "def"))
    --> =abc

    print(pcall(werr))
    --> ~false\t.*def

    print(pcall(werr))
    --> ~false\t.*cannot resume dead thread
end

-- Test coroutine.close() (5.4)
do
    print(pcall(coroutine.close))
    --> ~false\t.*value needed

    print(pcall(coroutine.close, 1))
    --> ~false\t.*must be a thread

    print(pcall(coroutine.close, coroutine.running()))
    --> ~false\t.*cannot close running thread

    local co = coroutine.create(function()
        coroutine.yield()
    end)

    coroutine.resume(co)
    print(coroutine.close(co))
    --> =true
    print(coroutine.status(co))
    --> =dead
    local co = coroutine.create(function()
        local t = {}
        setmetatable(t, {__close=function() error("ERR") end})
        local x <close> = t
        coroutine.yield()
    end)
    coroutine.resume(co)
    pcall(function() print(coroutine.close(co)) end)
    --> ~false\t.*ERR

    print(coroutine.status(co))
    --> =dead
end
